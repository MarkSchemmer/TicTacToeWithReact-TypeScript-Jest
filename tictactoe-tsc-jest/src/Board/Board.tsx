import * as React from 'react'
import Square from '../Square/Square'
import * as _ from '../Helpers/helpers'
interface IProps  {
    squares : Array<Array<_.Square>>
}

class Board extends React.Component<IProps> {
    render() {
        const { squares } = this.props
        return (
            <div>
                { squares.map((level) => level.map((sublevel) => {
                        return <Square
                                val={sublevel.val} 
                                coordinates={sublevel.coordinates}  /> 
                }))}          
            </div>
        )
    }
}

export default Board 