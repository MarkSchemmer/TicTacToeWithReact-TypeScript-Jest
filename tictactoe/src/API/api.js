

export const OPTIONS = {
    XWIN : 1,
    XLOSE : 2,
    TIE : 3, 
    MOREMOVES : 4
}


let threeInRow = (chr, [a,b,c], nb) =>  nb[a] === nb[b] && nb[b] === nb[c] && nb[a] === chr

export function IsWinner (board){

    let nb = board
    .reduce((acc,cur) => acc.concat(cur), [])
    .map( x => x.val )

    console.log(nb)

    let waysToWin = [
        [0,1,2],
        [0,4,8],
        [0,3,6],
        [1,4,7],
        [2,4,6],
        [2,5,8],
        [3,4,5],
        [6,7,8]
    ]

    for(let i = 0; i < waysToWin.length; i++) {

        console.log(threeInRow('X', waysToWin[i], nb))

        if(threeInRow('X', waysToWin[i], nb))
            return OPTIONS.XWIN
        else if (threeInRow('O', waysToWin[i], nb))
            return OPTIONS.XLOSE
    }

    return nb.every(x => x !== null ) ? OPTIONS.TIE : OPTIONS.MOREMOVES
}
