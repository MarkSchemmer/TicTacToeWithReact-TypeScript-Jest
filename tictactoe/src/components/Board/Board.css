
.row { 
    width:400px;
    height: 100px;
    /* border: 1px solid black; */
    margin: auto;
}


.column { 
    width: 130px;
    height: 100px;
    border : 1px solid black;
    display: inline-block;
}