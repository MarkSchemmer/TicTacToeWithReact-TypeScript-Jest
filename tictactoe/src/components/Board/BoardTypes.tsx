import React, { Component } from 'react'


interface ISquareAttributes {
    id : number, 
    coor : Array<number> 
    val : string | null 
}

interface IBoardWithMove {
    history : Array<Array<ISquareAttributes>>,
    Move : Array<Array<number>> | null,
    WhomMoved : string | null 
}


export class SquareAttributes implements ISquareAttributes {
    public id : number; 
    public coor : Array<number>;
    public val : string | null;

    constructor(id : number, coor : Array<number>, val : string | null = null){
        this.id = id
        this.coor = coor 
        this.val = val 
    }
}

export class BoardWithMoves implements IBoardWithMove {
    
    public history : Array<Array<SquareAttributes>>

    public Move : Array<Array<number>>

    public WhomMoved : string | null 
    
    constructor(history:Array<Array<SquareAttributes>>,
         Move : Array<Array<number>>, 
         WhomMoved : string | null = null ){
        this.history = history
        this.Move = Move 
        this.WhomMoved = WhomMoved
    }
}