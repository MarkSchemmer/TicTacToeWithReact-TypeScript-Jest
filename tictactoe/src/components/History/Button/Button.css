
.play-again {
    margin-top : 15px;
    padding: 10px;
}