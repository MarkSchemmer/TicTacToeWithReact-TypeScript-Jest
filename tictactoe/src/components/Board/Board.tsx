import React, { Component } from 'react'
import History from '../History/History'
import Square from '../Square/Square'
import './Board.css'

import * as api from '../../API/api'

import  * as BoardAttributes from './BoardTypes'

interface IProps {

}

interface IState {
    name : string, 
    board : Array<BoardAttributes.BoardWithMoves>,
    turn : number,
    IsWinner : number | null,
}

class Board extends Component<IProps, IState> {


    public state : IState = {
        name : 'Mark',
        board : [api.genBoard()],
        turn : 0,
        IsWinner : null, 
    }


    public onClick = (coor:Array<number>) => {
        const [x,y] = coor

        if(this.state.board[this.state.turn].history[x][y].val !== null) return  

        if(this.state.IsWinner !== null) return 

        // copy of history
        let copyOfHistory = this.state.board[this.state.turn].history

        // copy of Move 
        let copyOfMove = this.state.board[this.state.turn].Move 
        copyOfMove = [...copyOfMove, [x,y]]

        let whosMoving
    

        if(this.state.turn%2==0)
            whosMoving = 'X'
        else 
            whosMoving = 'O'

        copyOfHistory[x][y].val = whosMoving

        let newCopy = new BoardAttributes.BoardWithMoves(copyOfHistory, copyOfMove, whosMoving)

        this.setState((prevState) => {
            return {
                turn : prevState.turn+1,
                board : [...this.state.board, newCopy]
            }
        }, () => this.checkForWinner())
    }
 

    public checkForWinner = () => {
        console.log('the board: ',this.state.board)
        const hasWinner = api.IsWinner(this.state.board[this.state.turn].history)
        let whichOption = null 

        if(hasWinner===api.OPTIONS.XWIN)
            whichOption = api.OPTIONS.XWIN
        else if (hasWinner === api.OPTIONS.XLOSE) 
              whichOption = api.OPTIONS.XLOSE
        else if (hasWinner === api.OPTIONS.TIE)
                whichOption = api.OPTIONS.TIE

     if (whichOption)
        this.setState( { IsWinner : whichOption }, () => {
            this.setWinner()
        }) 
    }

    public setWinner = () => {
        if(this.state.IsWinner !== null){

            if(this.state.IsWinner===api.OPTIONS.TIE){
                return (
                    <React.Fragment>
                        TIE GAME!
                    </React.Fragment>
                )
            } else {
                if(this.state.IsWinner===api.OPTIONS.XWIN){
                    return (
                        <React.Fragment>
                            Player X You Win!
                        </React.Fragment>
                    )
                } else {
                   return ( <React.Fragment>
                        Player O You Win!
                   </React.Fragment> )
                }
            }

        } else { 
            return (
                <React.Fragment>
                    Player { this.state.turn % 2 === 0 ? <strong>{'X'}</strong> : <strong>{'O'}</strong> } your move!
                </React.Fragment>
            )
        }
    }

    public render () {
        return (
            <React.Fragment>
                <div className="main">
                <History history={this.state.board} /> 
                  <div className="greetings">Greetings { this.state.name }</div>
                  <div className="whoms-turn">  { this.setWinner() } </div>
                    <div className="board">
                            { this.state.board[this.state.turn].history.map((x:any) => 
                               <div className="row" key={ Date.now() * (Math.random() * 100) }>
                                    { x.map((z:any) => <Square key={z.id} {...z} 
                                        onClick={this.onClick.bind(this)} /> )}
                               </div> )}
                    </div>
                </div>
            </React.Fragment>
        )
    }
}

export default Board 

