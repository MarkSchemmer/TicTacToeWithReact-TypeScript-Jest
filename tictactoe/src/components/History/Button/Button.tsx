import React, { Component } from 'react'
import * as types from '../../Board/BoardTypes'
import './Button.css'


interface IProps {
    obj : types.BoardWithMoves, 
    index : number, 
    x : number, 
    y : number, 
    whoIsClicked : number | null, 
    handleClick : any 
}

class Button extends Component<IProps> {

    constructor(props : IProps){
        super(props)
        this.toggleIsClicked = this.toggleIsClicked.bind(this)
    }


    public toggleIsClicked = () : void => {
        this.props.handleClick(this.props.index)
    }

    
    public genButton = (content : string, shouldHighlight : boolean) => 
        shouldHighlight ? <button onClick={ this.toggleIsClicked }><strong>{ content }</strong></button> 
                        : <button onClick={ this.toggleIsClicked }> {content} </button>  

    public outputingButton = (obj : types.BoardWithMoves, 
                              index : number, 
                              x : number, 
                              y : number) => {

        let complexButton = ` Player: ${obj.WhomMoved} Move :${index} Coordinate : (${x}, ${y}) `
        let originalButton

        switch(index){
            case 0 :
                originalButton = (
                    <React.Fragment key={index}>
                        { this.genButton("Start of Match", this.props.whoIsClicked===index) } 
                    </React.Fragment>
                )
            break 

           default :
                    originalButton = (
                    <React.Fragment key={index*Date.now()}>
                            { this.genButton(complexButton, this.props.whoIsClicked===index) }
                    </React.Fragment>
                    )
            break 
        }

        return originalButton
    }

    public render () {
       let { obj, index, x, y} = this.props
        return (
            <React.Fragment>
                { this.outputingButton(obj, index, x, y)}
            </React.Fragment>
        )
    }
}


export default Button 