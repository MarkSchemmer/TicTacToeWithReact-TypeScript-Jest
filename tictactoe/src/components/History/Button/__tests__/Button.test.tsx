import * as React from 'react'
import * as enzyme from 'enzyme'
import  Adapter from 'enzyme-adapter-react-16'
import renderer from 'react-test-renderer'

import Button from '../Button'

import * as types from '../../../Board/BoardTypes'

import * as api from '../../../../API/api.js'

 
enzyme.configure({ adapter: new Adapter() });


describe(' dummy for Button ', () => {
    it(' dummy it test', () => {
        expect(true).toEqual(true)
    })
})



/*
    create fakes props for tests
        - Does function get called when button clicked 
        - Test both outputs, such as index === 0 and output for index > 0 
*/

describe(' Testing <Button /> with fake props ', () => {

    let buttonFakeProps, handleClick

    beforeEach(() => {

        handleClick = jest.fn()

        buttonFakeProps = {
            obj : api.genBoard(),
            index : 1,
            x : 0,
            y : 0, 
            whoIsClicked :  1,
        }
    })

    it(' will handle click be called ', () => {
        const wrapper = enzyme.shallow( 
            <Button 
                {...buttonFakeProps} 
                handleClick={handleClick} /> )
        wrapper.find('button').simulate('click')
        wrapper.update()

        expect(handleClick).toHaveBeenCalled() 
    })

    it(' what is the output of click ', () => {
        const wrapper = enzyme.shallow(
            <Button 
                {...buttonFakeProps}
                handleClick={handleClick} /> )

         wrapper.find('button').simulate('click')
         wrapper.update()

        const { obj, index, x, y, whoIsClicked } = buttonFakeProps

        
         expect(wrapper.find('button').text()).toEqual("  Player: null Move :undefined Coordinate : (0, 0)  ")
        
    })
})



// snapshot testing 



describe(' snapshot test for <Button /> ', () => {


    let buttonFakeProps, handleClick

    beforeAll(() => {

        handleClick = jest.fn()

        buttonFakeProps = {
            obj : api.genBoard(),
            index : 1,
            x : 0,
            y : 0, 
            whoIsClicked :  null,
        }
    })

    it(' snapshot test', () => {
        let handleClick = jest.fn()
        const tree = renderer.create( <Button 
                {...buttonFakeProps} 
                handleClick={handleClick} /> ).toJSON()

        expect(tree).toMatchSnapshot() 
    })
})