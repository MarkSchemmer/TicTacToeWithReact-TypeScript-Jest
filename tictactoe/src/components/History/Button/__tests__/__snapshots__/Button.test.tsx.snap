// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` snapshot test for <Button />   snapshot test 1`] = `
<button
  className="hist-btn"
  onClick={[Function]}
>
   
   Player: null Move :1 Coordinate : (0, 0) 
   
</button>
`;
